website/blog/2023-12-08-silver.md

```diff
@@ -0,0 +1,585 @@
+This means that most new releases will focus on stability instead of adding new
+features, unless those features contribute to making Metals more relaible.
+
+<table>
+<tbody>
+  <tr>
+    <td>Commits since last release</td>
+    <td align="center">144</td>
+  </tr>
+  <tr>
+    <td>Merged PRs</td>
+    <td align="center">110</td>
+  </tr>
+    <tr>
+    <td>Contributors</td>
+    <td align="center">16</td>
+  </tr>
+  <tr>
+    <td>Closed issues</td>
+    <td align="center">41</td>
+  </tr>
+  <tr>
+    <td>New features</td>
+    <td align="center">5</td>
+  </tr>
+</tbody>
+</table>
+
+For full details:
+[https://github.com/scalameta/metals/milestone/62?closed=1](https://github.com/scalameta/metals/milestone/62?closed=1)
+
+Metals is a language server for Scala that works with VS Code, Vim, Emacs and
+Sublime Text. Metals is developed at the [Scala Center](https://scala.epfl.ch/)
+and [VirtusLab](https://virtuslab.com) with the help from contributors from the
+community.
+
+## TL;DR
+
+Check out [https://scalameta.org/metals/](https://scalameta.org/metals/), and
+give Metals a try!
+
+- [Use presentation compiler for synthetic decorations](#use-presentation-compiler-for-synthetic-decorations)
+- [Base Metals view on indexing information](#base-metals-view-on-indexing-information)
+- [Go to definition for scaladoc](#go-to-definition-for-scaladoc)
+- [Custom project root setting](#custom-project-root-setting)
+- [Bsp errors in status](#bsp-errors-in-status)
+- [Add back unintentionally removed support for 2.11.x](#add-back-unintentionally-removed-support-for-211x)
+- [Changes for plugin authors](#changes-for-plugin-authors)
+
+## Use presentation compiler for synthetic decorations
+
+Up until this release Metals would use the semanticdb information for displaying
+synthetic decorations such as inferred types or implicit conversions. This was
+not ideal as it meant that the information was not always up to date and could
+be inconsistent with the current state of code.
+
+![synthetics](https://i.imgur.com/td11vRH.gif)
+
+To fix that [jkciesluk](https://github.com/jkciesluk) switched to using the
+presentation compiler for this information. This means that each time code
+changes the decorations will be refreshed. This also added a number of new
+decorations that were not previously available for Scala 3 and it will be on par
+with the Scala 2 ones.
+
+## Base Metals view on indexing information
+
+Metals tree view enables users to navigate their workspace in a more structured
+way by directly showing the build modules and their packages, classes, objects
+etc. It would also allow users to navigate their dependency jars. Previously,
+this was based on special Scala 2 signature information, which means it wouldn't
+work in Scala 3.
+
+From this release we changed it to use the same information that is used for
+indexing in case of jars and semanticdb for local workspace files. This means
```

Comment:
If this uses the same information, does this also improve the performance? If so, it might be nice to add that as well.

Reply:
It's very similar I think in terms of performance (maybe a bit faster), we do need to index it twice, since the information we need is totally opposite than what we need from normal indexes. And I didn't want to make global indexes much larger. Though here we use the same fast indexing mechanism instead of more costly loading of classfiles previously.

With normal indexes we need `symbol -> file` and here we need `file -> symbol`. Creating that info from existing indexes was slow since we needed to search the whole map of dependencies for symbols in a particular file. If we saved that in memory, we would have a double the memory usage for indexes.

Reply:
Checking again, the performance might sometimes be a bit worse, but it's only noticeable in larger jars (and still not much slower). I can make it faster but the most important thing here was making sure it always work reliably. Also, I think the biggest issue with performance is not actually connected to the method of indexing.

I will take a look at it at some point.